README.md
# Distance

[![CI](https://github.com/pkinney/distance/actions/workflows/ci.yaml/badge.svg)](https://github.com/pkinney/distance/actions/workflows/ci.yaml)
[![Hex.pm](https://img.shields.io/hexpm/v/distance.svg)](https://hex.pm/packages/distance)

Provides a set of distance functions for use in GIS or graphic applications.

## Installation

```elixir
defp deps do
  [{:distance, "~> 1.0"}]
end
```

## Functions

### Point-Point Distance

Calculate geometric distance between two or more points (two- or three-dimensional):

```elixir
Distance.distance({2, -1}, {-1, 3}) # => 5
Distance.distance({2, -1, 4}, {-1, 3, 2}) # => 5.385...

Distance.distance([{2.5, 2.5}, {4, 0.8}, {2.5, 3.1}, {2.5, 3.1}]) # => 5.013...
```

Calculate the square of the geometric distance between two points (useful as
  a faster way to compare distances between points without the need for an
  expensive square root operation):

```elixir
Distance.distance_squared({2, -1}, {-1, 3}) # => 25
Distance.distance_squared({2, -1, 4}, {-1, 3, 2}) # => 29
```

### Point-Segment Distance

Calculate geometric distance between a point and the closest point on a line
segment.  For instance the distance between the point (3, 3) and the line
segment between (-2, 1) and (5, 3).

```elixir
Distance.segment_distance({3, 2}, {-2, 1}, {5, 3}) # => 0.412...
```

Similar to the distance function, there is a squared version for faster
calculations when needed:

```elixir
Distance.segment_distance_squared({3, 2}, {-2, 1}, {5, 3}) # => 0.170...
```

### Angular Functions

Get bearing from one point to another and project from one point a given
distance along a direction.  Additionally, there are useful methods for 
simplifying angles to coterminal angles and measuring differences between
directions.

Angles are measured as radians off of the positive x-axis in the direction of
the positive y-axis.


### Great Circle Distance

Calculate great circle distances (shortest travel distance on the surface of
a spherical Earth) given a two longitude-latitude pairs.  This is an implementation
of the [Haversine formula](https://en.wikipedia.org/wiki/Haversine_formula)
and approximates using a spherical (non-ellipsoid) Earth with a
mean radius of 6,371,008.8 meters derived from the WGS84 datum.

The function accepts two tuples in the form of `{longitude, latitude}` and
returns the distance in meters. It will also accept a List of tuples.

```elixir
Distance.GreatCircle.distance({-96.796667, 32.775833}, {126.967583, 37.566776}) # => 10974882.74...

Distance.GreatCircle.distance([
  {-96.796667, 32.775833},
  {126.967583, 37.566776},
  {151.215158, -33.857406},
  {55.274180, 25.197229},
  {6.942661, 50.334057},
  {-97.635926, 30.134442}
]) # => 44728827.849...
```


### Vincenty's Inverse Formula Distance

Calculate distance per [Vincenty's inverse formula](https://en.wikipedia.org/wiki/Vincenty%27s_formulae)
(shortest travel distance on the surface of an [oblate spheroid](https://en.wikipedia.org/wiki/Spheroid#Oblate_spheroids) Earth) given two longitude-latitude pairs.

This method is iterative and more costly than other methods, such as the [great circle](lib/distance/great_circle.ex) method, but also potentially more accurate. 
It is important to note that [nearly antipodal points](https://en.wikipedia.org/wiki/Vincenty%27s_formulae#Nearly_antipodal_points) can cause convergence issues with this method.

The function accepts two tuples in the form of `{longitude, latitude}` and
returns the distance in meters. It will also accept a List of tuples.

```elixir
Distance.Vincenty.distance({-96.796667, 32.775833}, {126.967583, 37.566776}) # => 10997423.55...

Distance.Vincenty.distance([
  {-96.796667, 32.775833},
  {126.967583, 37.566776},
  {151.215158, -33.857406},
  {55.274180, 25.197229},
  {6.942661, 50.334057},
  {-97.635926, 30.134442}
]) # => 44737835.514...
```


# Contributing

In order to run all of the checks locally, use the `validate` mix task:

```bash
mix validate
mix test
```


CHANGELOG.md
# Change Log


## [1.1.1] - 2023-05-30

### Changed

- Loosen `seg_seg` dependency requirement to include `0.1.x` and `1.x`


## [1.1.0] - 2022-05-02

### Added

- Add Vincenty's Inverse Formula Distance (thanks @Cameron-Kurth)


## [1.0.0] - 2021-11-24

### Added

- `project/3` - Projects a given distance in a given direction away from a point
- `angle_to/2` - Returns direction from one point to another
- `min_coterminal_angle/1` - Coterminal angle closest to 0 for the given angle
- `min_positive_coterminal_angle/1` - Smallest positive coterminal angle for the given angle
- `angular_difference/2` - Angular difference between two angles

### Changed

- Update CI to use Github Actions instead of travis-ci
- Update dependencies


lib/distance/great_circle.ex
defmodule Distance.GreatCircle do
  @moduledoc ~S"""
  Calculate great circle distances (shortest travel distance on the surface of
  a spherical Earth) given a two longitude-latitude pairs.  This is an implementation
  of the [Haversine formula](https://en.wikipedia.org/wiki/Haversine_formula)
  and approximates using a spherical (non-ellipsoid) Earth with a
  mean radius of 6,371,008.8 meters derived from the WGS84 datum.

  The function accepts two tuples in the form of `{longitude, latitude}` and
  returns the distance in meters. It will also accept a List of tuples.
  """

  @type coords() :: {number(), number()}

  @pi_over_180 3.14159265359 / 180.0
  @radius_of_earth_meters 6_371_008.8

  @doc """
  Returns the great circle distance in meters between two points in the form of
  `{longitude, latitude}`.

  ## Examples
      iex> Distance.GreatCircle.distance({-105.343, 39.984}, {-105.534, 39.123})
      97129.22118968463
      iex> Distance.GreatCircle.distance({-74.00597, 40.71427}, {-70.56656, -33.42628})
      8251609.780265334
  """

  @spec distance(coords(), coords()) :: float()
  def distance({lon1, lat1}, {lon2, lat2}) do
    a = :math.sin((lat2 - lat1) * @pi_over_180 / 2)
    b = :math.sin((lon2 - lon1) * @pi_over_180 / 2)

    s = a * a + b * b * :math.cos(lat1 * @pi_over_180) * :math.cos(lat2 * @pi_over_180)
    2 * :math.atan2(:math.sqrt(s), :math.sqrt(1 - s)) * @radius_of_earth_meters
  end

  @doc """
  Returns the great circle distance in meters along a linestring defined by the
  List of `{longitude, latitude}` pairs.

  ## Examples
      iex> Distance.GreatCircle.distance([
      ...>  {-96.796667, 32.775833},
      ...>  {126.967583, 37.566776},
      ...>  {151.215158, -33.857406},
      ...>  {55.274180, 25.197229},
      ...>  {6.942661, 50.334057},
      ...>  {-97.635926, 30.134442}])
      44728827.84910666
  """
  @spec distance(list(coords())) :: float()
  def distance([]), do: 0.0
  def distance([_]), do: 0.0

  def distance([p1, p2 | tail]) do
    distance(p1, p2) + distance([p2 | tail])
  end
end
